Sprint_6

Vezaris opened the pull request:
Написаны тесты для классов Lion, Feline и Cat.
В тестах есть параметризация и моки.

Vezaris committed:
Первый коммит.
Написаны тесты по заданию.

egorvas commented on src/test/java/ru/yandex/practicum/CatTest.java:
```diff
@@ -0,0 +1,38 @@
+    public void getFoodForPredator2() throws Exception {
+        cat = new Cat(felineMock);
+        List <String> expected = List.of("Животные", "Птицы", "Рыба");
+        Mockito.when(felineMock.eatMeat()).thenReturn(List.of("Животные", "Птицы", "Рыба"));
```
**Можно лучше:** у тебя List.of("Животные", "Птицы", "Рыба") вынесено в переменную. Можно переиспользовать.

egorvas commented on src/test/java/ru/yandex/practicum/FelineTest.java:
```diff
@@ -0,0 +1,79 @@
+        assertEquals(expectedFoodForPredator, actualFoodForPredator);
+    }
+    @Test // Метод getFood получает список еды для хищника, травоядного и неизвестного значения
+    public void getFoodReturnFood() throws Exception {
```
**Нужно исправить:** Тут довольно грубая ошибка. В случае если произойдет неожиданный эксепшен в позитивном сценарии, то твой тест не упадет попав в catch.
В случае отсутствия эксепшена при негативном сценарии, твой тест также не упадет, так как у тебя вообще нет ассерта на это.
Тут нужно разделить на два теста - параметризованный с позитивными сценариями и отдельный на эксепшен. Причем нужно обязательно проверять, что эксепшен произошел.

egorvas commented on src/test/java/ru/yandex/practicum/LionTest.java:
```diff
@@ -0,0 +1,80 @@
+        assertEquals(expected, actual);
+    }
+    @Test // Возвращается корректное значение hasMane в зависимости от пола
+    public void doesHaveMane() throws Exception {
```
**Нужно исправить:** аналогично замечанию выше

Vezaris committed:
Переделан конструктор для класса Lion
Добавлена проверка на схождение текста ошибки при неизвестном значении в праметрезированном тесте.
Добавлены обработки ошибок:
- неизвестная ошибка в позитивном сценарии
- отсутствие ошибки при негативном сценарии

egorvas commented on src/test/java/ru/yandex/practicum/FelineTest.java:
```diff
@@ -0,0 +1,83 @@
+    @Spy
+    Feline felineSpy;
+    String animalType;
+    final List<String> EXPECTEDFOODFORPREDATOR = List.of("Животные", "Птицы", "Рыба");
```
**Нужно исправить:** используй нижнее подчеркивание чтобы разделить слова в константах

egorvas commented on src/test/java/ru/yandex/practicum/FelineTest.java:
```diff
@@ -0,0 +1,83 @@
+        String expectedMessageErrorType = "Неизвестный вид животного, используйте значение Травоядное или Хищник";
+        try {
+            List<String> actualFood = felineSpy.getFood(animalType);
+            if ("Травоядное".equals(animalType)) {
```
**Нужно исправить:** если тебе нужно в параметризированный тест запихнуть кучу if/else чтобы описать логику, значит это плохое место для параметризированного теста.

egorvas commented on src/test/java/ru/yandex/practicum/FelineTest.java:
```diff
@@ -0,0 +1,83 @@
+            }  else {
+                throw new Exception ("Отсутствует ошибка: " + expectedMessageErrorType);
+            }
+        } catch (Exception exception) {
```
**Нужно исправить:** у тебя по-прежнему полуается, что если произойдет неожиданный эксепшен в позитивном сценарии то твой тест не упадет.

Vezaris replied:
@egorvas а можно получить пример кода, что бы тест упал? У меня при ошибке в позитивном сценарии, тест все отлавливает.

egorvas replied:
Это касалось старой реализации, сейчас ок

egorvas commented on src/test/java/ru/yandex/practicum/FelineTest.java:
```diff
@@ -0,0 +1,83 @@
+            }
+        } catch (Exception exception) {
+            if (exception.getMessage().equals(expectedMessageErrorType)) {
+            } else {
```
**Нужно исправить:** зачем тут конструкция if с пустым условием? Ты мог написать
`if (!exception.getMessage().equals(expectedMessageErrorType))`
И тогда не пришлось бы делать else вообще

Vezaris committed:
Изменил тесты, которые использовали много if/else.
Разделены тесты для позитивного и негативного сценария.
Проверены сценарии получения неожиданных ошибок, или отсутствие ошибки в негативном сценарии. Теперь, если в негативном сценарии не рпидет ожидаемая ошибка, то будет ошибка на отсутствие ошибки.

egorvas commented on src/test/java/ru/yandex/practicum/LionParamTest.java:
```diff
@@ -0,0 +1,93 @@
+        String expectedErrorMessage = "Используйте допустимые значения пола животного - самей или самка";
+        try {
+            lion = new Lion(felineMock, "Другое");
+            throw new Exception("Отсутствует ожидаемая ошибка");
```
**Можно лучше:** лучше всегда использовать ассерт, чем каст эксепшена. Как минимум это лучше смотрится в отчетах.
